_posts/2020-10-22-better-structure-for-kotlin-projects.md

```diff
@@ -0,0 +1,125 @@
+anymore as we don't have to carry Java code anywhere in Kotlin projects.
+
+Yet, we still utilize projects structures like this:
+
+```
+src
+├── main                   
+│   └── kotlin
+│       └── com
+│           └── companyname
+│               └── projectname
+│                   ├── controllers
+│                   ├── services 
+│                   └── Main.kt
+├── test                   
+│   └── kotlin
+│       └── com
+│           └── companyname
+│               └── projectname
+│                   ├── controllers
+│                   ├── services 
+│                   └── MainTest.kt
+```
+Not only is it complex and verbose, but it also introduces other issues.
+If you expand the project structure in an IDE it consumes more width
+on the screen leaving less space for the editor. The same issue
+appears during code review, especially if you use the side-by-side view.
+Long paths, like 
+`src/main/kotlin/com/companyname/projectname/services/UserService.kt`,
+often don't fit into the screen. For instance, GitHub
+might show it like `src/main/kotlin/com/companyname/projectname/servi...`.
+You might say that we all work from home and have 27" monitors, and you
+will be right. Yet, many use only laptops, where this
+issue gets more important. In any way, why would you want to carry
+boilerplate in your projects? Is Kotlin not about being pragmatic?
+Also, people who only start learning Kotlin and have no
+Java background might find it confusing as well. Even if you just need
+to take a glance at a project you would have to go through numerous
+nested folders which is not the best experience.
+
+What is the solution? Instead, we could use a structure like this
+
+```sh
+src
+├── controllers
+├── services 
+└── Main.kt
+testSrc                   
+├── controllers
+├── services 
+└── MainTest.kt
+```
+
+No, I do not suggest you remove package declarations. Kotlin allows
+declaring classes/functions whose packages do not match the folder structure.
+
+The following code will get compiled without issues if it is placed
+in the root of a source set.
+
+```kotlin
+package com.companyname.projectname
+
+fun main() {
+    println("Hello!")
+}
+```
+My suggestion is to omit folders corresponding to the root package,
+and use a single directory per subpackage.
+```sh
+com.companyname.projectname ->             src/
+com.companyname.projectname.services ->    src/services
+com.companyname.projectname.controllers -> src/controllers
+
+```
+Next, how do we get rid of `src/main/kotlin` and `src/test/kotlin`?
+This could be achieved by overriding `sourceSets` in a Gradle project.
+Just add the following instruction to your `build.gradle`.
+```
+sourceSets {
+    main.kotlin.srcDirs = ['src']
+    test.kotlin.srcDirs = ['testSrc']
+}
+```
+
+The same could be done for the `resources` folder.
+
+These simple manipulations allow us to reduce the amount of boilerplate
+in our code bases. Of course, it might not fit all types of projects.
```

Comment:
Could probably combine the last two sentences to wrap it all up nicely... "Of course, it might not fit all types of projects, but in the case of small code bases it comes in really handy."